Humanoids w/ Top-level Darkview or XRAY Vision have Glowing Eyes

KasparoVy opened the pull request:
:cl:
rscadd: The eyes of humanoids who have XRAY vision or top-level darkview will appear to shine in the dark.
/:cl:

The eyes are hidden behind anything that would normally cover them.
The mob's icons update when the appropriate genetic traits are toggled to show this feature.

![foxglowyeyes](https://user-images.githubusercontent.com/8165455/34345234-67b25d72-e9bb-11e7-95e2-c020606e93e8.jpg)

![glowindarkeyes](https://user-images.githubusercontent.com/12377767/34448893-246485d8-ecc0-11e7-81f9-affdd6171420.png)

tigercat2000 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1668,6 +1668,13 @@
 	G.synch()
 
+/mob/living/carbon/human/proc/eyes_glow() /*Used to check if eyes should glow in the dark. Returns the image of the eyes on the layer where they will appear to glow.
+											Eyes need to have significantly high darksight to glow unless the mob has the XRAY vision mutation. Eyes will not glow if they are covered in any way.*/
```
With this long of a comment, it should just be on the lines before the proc.

tigercat2000 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1668,6 +1668,13 @@
+/mob/living/carbon/human/proc/eyes_glow() /*Used to check if eyes should glow in the dark. Returns the image of the eyes on the layer where they will appear to glow.
+											Eyes need to have significantly high darksight to glow unless the mob has the XRAY vision mutation. Eyes will not glow if they are covered in any way.*/
+	var/obj/item/organ/internal/eyes/eyes = get_int_organ(/obj/item/organ/internal/eyes)
+	var/const/glow_threshold = 6
```
This should be a #define, not a constant defined in the function.

tigercat2000 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1668,6 +1668,13 @@
 	G.synch()
 
+/mob/living/carbon/human/proc/eyes_shine() /*Used to check if eyes should shine in the dark. Returns the image of the eyes on the layer where they will appear to shine.
+											Eyes need to have significantly high darksight to shine unless the mob has the XRAY vision mutation. Eyes will not shine if they are covered in any way.*/
```
This comment should be before the proc, since it's so long.

FreeStylaLT commented:
This is a glowing example of a cool PR

tigercat2000 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1668,6 +1668,13 @@
+/mob/living/carbon/human/proc/eyes_shine() /*Used to check if eyes should shine in the dark. Returns the image of the eyes on the layer where they will appear to shine.
+											Eyes need to have significantly high darksight to shine unless the mob has the XRAY vision mutation. Eyes will not shine if they are covered in any way.*/
+	var/obj/item/organ/internal/eyes/eyes = get_int_organ(/obj/item/organ/internal/eyes)
+	var/const/shine_threshold = 6
```
This should be a #define, not a constant, and especially not an inline constant.

KasparoVy replied:
Which is preferable here? A define limited to this scope that is undefined at the end of the proc or one that is put in __DEFINES\mob.dm and left?

tigercat2000 replied:
The latter.

tigercat2000 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1668,6 +1668,13 @@
+											Eyes need to have significantly high darksight to shine unless the mob has the XRAY vision mutation. Eyes will not shine if they are covered in any way.*/
+	var/obj/item/organ/internal/eyes/eyes = get_int_organ(/obj/item/organ/internal/eyes)
+	var/const/shine_threshold = 6
+	if(istype(eyes) && ((eyes.get_dark_view() > shine_threshold) || (XRAY in mutations)) && get_location_accessible(src, "eyes"))
```
Seperate this into more conditionals please.
```
if(!istype(eyes))
    return null
if(!(eyes.get_dark_view() > SHINE_THRESHOLD) || !(XRAY in mutations))
    return null
if(!get_location_accessible(src, "eyes"))
    return null

tigercat2000 commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -160,6 +160,7 @@ Please contact me on #coderbus IRC. ~Carn x
 		overlays += (new_overlays - old_overlays)
 		overlays -= (old_overlays - new_overlays)
 		cached_standing_overlays = new_overlays
+		overlays |= eyes_shine() //Only applies eyeglow if the eyes are uncovered and have significantly high darksight or XRAY vision.
```
This is an absolutely terrible way to do this, you should use what's already there and apply/remove the overlays correctly. This also means that the eyes_shine() proc should return a boolean value, not an image.

tigercat2000 commented on code/modules/surgery/organs/organ_internal.dm:
```diff
@@ -305,6 +305,13 @@
 	dna.write_eyes_attributes(src)
 
+/obj/item/organ/internal/eyes/proc/generate_icon(var/mob/living/carbon/human/H)
+	var/mob/living/carbon/human/G = H ? H : owner
```
While more compact, this is less readable than doing the following
```
/obj/item/organ/internal/eyes/proc/generate_icon(mob/living/carbon/human/HA)
var/mob/living/carbon/human/H = HA
if(!istype(H))
   H = owner
var/icon...

tigercat2000 commented on code/modules/surgery/organs/organ_internal.dm:
```diff
@@ -305,6 +305,13 @@
+/obj/item/organ/internal/eyes/proc/generate_icon(var/mob/living/carbon/human/H)
+	var/mob/living/carbon/human/G = H ? H : owner
+	if(istype(G))
+		var/icon/eyes_icon = new/icon('icons/mob/human_face.dmi', G.species.eyes)
```
`new /icon`

tigercat2000 commented on code/modules/surgery/organs/organ_internal.dm:
```diff
@@ -305,6 +305,13 @@
+	if(istype(G))
+		var/icon/eyes_icon = new/icon('icons/mob/human_face.dmi', G.species.eyes)
+		eyes_icon.Blend(eye_colour, ICON_ADD)
+		return(eyes_icon)
```
`return eyes_icon`

tigercat2000 commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -160,6 +160,7 @@ Please contact me on #coderbus IRC. ~Carn x
 		overlays += (new_overlays - old_overlays)
 		overlays -= (old_overlays - new_overlays)
 		cached_standing_overlays = new_overlays
+		overlays |= eyes_shine() //Only applies eyeshine if the eyes are uncovered and have significantly high darksight or XRAY vision.
```
This is an absolutely terrible way to do this, you should use what's already there and apply/remove the overlays correctly. This also means that the eyes_shine() proc should return a boolean value, not an image.

KasparoVy replied:
The image's .layer needs to be lighting_layer + 1. This needed to dodge the above code else it'd get shuffled into a different layer. There is no code that currently accomplishes this but I'll keep looking. It's clear I'll have to be a bit more creative.

Aurorablade commented:
We need to trick someone into updating our icon code to tgs

Aurorablade commented:
Or a varient of

Aurorablade commented:
also i cann;t think of this without thinking of super mario sunshine.

KasparoVy commented:
Now I can't get the OST out of my head. Thanks! :laughing:

@tigercat2000 I've made changes addressing your review. [This code here](https://github.com/ParadiseSS13/Paradise/pull/8489/files#diff-fddcb82e477f30bac1d57ac11937ae0dR1332) is more/less placeholder while I prompt for your thoughts on how this is handled. I can further abstract this by developing a light 'misc appearance effects' system.

_Eyes covered by hair can still shine... Just not through the hair (anymore)!_
![xrayeyesindarkbangscover](https://user-images.githubusercontent.com/12377767/34451555-70dc3f8c-ecf7-11e7-9646-129bf3cb3b9d.PNG)

Aurorablade commented:
what i was gonna do before i gave up on our icon code (OH GOD SOMEONE UPDATE IT) was mage it a racial var for species, like has eyeshine or something then check it.

KasparoVy commented:
I thought about that too but then I figured it'd probably be best to keep it tied to darkview and xray since it has the added functionality of persisting if you transplant the eyes into a different mob without requiring more backend

Aurorablade commented:
actually i am gonna correct myself, I was gonna attack the var to the organ itself. then run the check based on the value on the organ. It getting to be that time of night my brain stops being coherent on some things but not others.

davidchan13 commented:
Questions:

Does this strictly work off the organ or does the implant version as well?

Does this affect Thermal vision for those that get the implant?

Does this affect vampires(who get X-ray with true form for no apparent reason)

KasparoVy commented:
1. It uses the primary vision organ for the dark_view assessment and checks the mob's mutations for XRAY vision.
2. No this feature does not interact with thermal vision.
3. If vampires get the XRAY vision `mutation` then yes it would affect them. I will review this.

davidchan13 commented:
I guess I'm unclear if the implant does mutation or something else, and I'm not sure what the code does for vampires.

KasparoVy commented:
The implants seem to do something else outside of the mutation so it'll require different checks that I haven't implemented, and Vampires have something snowflaked into species.dm [here](https://github.com/ParadiseSS13/Paradise/blob/c799ed6e429c27bc7caba4c0e5627c6b5e7ab230/code/modules/mob/living/carbon/human/species/species.dm#L498).

The goal of this PR is to implement eyeshine for organic eyes in humanoid mobs that have either high enough darksight or the XRAY vision mutation. I'm going to draw the line there and say that shining cybernetic or vampire eyes will needs its own PR

Dumbdumn5 commented:
This looks so fucking cool.

davidchan13 commented:
@KasparoVy Thank you for answering my questions! Definitely could agree for non organics to get eyeshine in the not-to-distant future. Personally I'd push to go as far as Mesons having their own shine effect, but that's something for another PR, another day.

SamHPurp commented:
> Eyes covered by hair can still shine... Just not through the hair (anymore)!

@KasparoVy is buffing emos! 😆 

This is a neat idea, @KasparoVy

Aurorablade commented:
We may need to address the vampire issue here due to mAh valids issues. Why I had thought to tie the value to the eye organ, but don't know how to handle xray then.

Quantum-M commented:
The color of the eye shine is dependent on the character's eye color, right?
If so, what about the eye shine of characters with dark eye colors (e.g. black)?

Aurorablade commented:
hey hey kasp...kasp...what do we have to do to bribe you to improve some of our icon code overall..cuase i know some stuff we NEED...

KasparoVy commented:
Alright I checked up on the vampire thing and they don't suddenly get shining eyes when their full power kicks in.

Vivalas commented:
This is neat but I need to play devil's advocate here... are there any hairstyles that completely cover eyes? Again like colorblindness this is cool on its own, but it is also nice for balance and long hair would nullify that.

Aurorablade commented:
no full cover but that would be a player choice issue there.

KasparoVy commented:
@Vivalas Not on Tajara and Vulpkanin. Tajara have a style that covers one eye.

Humans might, though. That'll only make a difference if they get XRAY vision mutation.

Allfd commented:
Time to add long hair gives chance to trip when running?

#humansarethereal❄️s

davidchan13 commented:
@Allfd long hair and floor braids make all grab attempts instant agressive for free

davidchan13 commented:
@Allfd Make long hair/floor braids automatically upgrade grabs to aggressive with a faster upgrade to neck grabs on top of that.

KasparoVy commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -1324,6 +1329,14 @@ var/global/list/damage_icon_parts = list()
+	misc_effect_overlays.Cut()
+
+	//Begin appending miscellaneous effects.
+	if(eyes_shine())
```
This code here is more/less placeholder while I prompt for your thoughts on how this is handled @tigercat2000. I can further abstract this by developing a light datun-based 'misc appearance effects' system so this update_misc_effects() can check a mob's misc_appearance_effects list that would have a reference to an eye_shine_effect datum thing that would have a proc called .assess() or something that would call the below checks and return the overlay.

Aurorablade replied:
i personally think it would be interesting espically if we could kinda expand it further to do limb icon generation like TG does, or if it makes applying the debrained sprite i never got to work easier. by that i mean https://github.com/tgstation/tgstation/blob/bc20a758d5f2e75e93587a58c4cd0ea5009bf243/code/modules/mob/living/carbon/update_icons.dm#L254  and https://github.com/tgstation/tgstation/blob/bc20a758d5f2e75e93587a58c4cd0ea5009bf243/code/modules/surgery/bodyparts/bodyparts.dm#L208 but up to tiger (our icon code is bad and should feel bad)

tigercat2000 replied:
It looks a lot better now, thank you.

KingPhilipIII commented:
NV goggles should produce this as well.

KasparoVy commented:
Another PR, this is for organic eyes.

KingPhilipIII commented:
Fair enough. This whole darksight debacle has opened a can of possible balance changes for NV stuff.

SamHPurp commented:
`but its balance implications wouldn't do this any favours.`

🙄

KasparoVy commented:
No it doesn't check for implants. If I do it for implants I'd do the eyeglow for implants and NV goggles in a separate PR

Vivalas commented:
I think combined with the other PR that just got merged increasing flash damage, the decoupling PR's balance detriments would be mitigated.

tigercat2000 commented on code/__HELPERS/icons.dm:
```diff
@@ -897,3 +897,12 @@ proc/rand_hex_color()
+	if(istype(main) && istype(mask))
+		mask.Blend(rgb(255,255,255), ICON_SUBTRACT) //Make all pixels on the mask as black as possible.
+		mask.Opaque(rgb(255,255,255)) //Make the transparent pixels (background) white.
+		mask.BecomeAlphaMask() //Make all the black pixels vanish (fully transparent), leaving only the white pixels.
```
I never knew this was a thing in the iconprocs library and it could probably be used on the gibber now that I know about it.

KasparoVy replied:
Yeah,  I really busted my brain trying to figure out how to write this proc and BecomeAlphaMask & the other icon helpers made me jump for joy. I think there's a miriad of places this one can be used.

KasparoVy commented:
Note to self: https://github.com/PolarisSS13/Polaris/pull/4627

Arokha beat me and I might have just wasted a bunch of time (lol, as if it mattered) figuring out that get_icon_difference proc.

Allfd commented:
@kasparoVV

So is this a Polaris port now or just a change to this PR?

KasparoVy commented:
@Allfd No idea what to call it at this point. I'm still on a train so I can't evaluate whether the above mentioned PR's method of doing this is relevant to us.

KasparoVy commented:
Yeah, the methods in that other version of this PR don't apply to this codebase.

I refactored this anyway and gave the PR a bit more clout by addressing the suggestions for synthetic eyes & cybernetic eye implants to cause eyeshine.

KasparoVy commented:
That's all I've got for this PR unless changes are requested on further review. I'll do goggles and such in another one.
